paccmann_generator/reinforce_sets.py

```diff
@@ -0,0 +1,485 @@
+
+        smiles_tensor = torch.cat(smiles_num, dim=0)
+        if target == 'efficacy':
+            smiles_tensor = smiles_tensor.narrow(1, len(smiles_tensor[0]) - 250, 250)
+
+        return smiles_tensor
+
+    def get_reward_affinity(self, valid_smiles: List, protein: str) -> np.array:
+        """Get the reward from affinity predictor.
+        
+        Args:
+            valid_smiles (List): A list of valid SMILES strings.
+            protein (str): Name of target protein
+
+        Returns:
+            np.array: computed reward (fixed to 1/(1+exp(x))).
+        """
+
+        smiles_tensor = self.smiles_to_numerical(valid_smiles, target='affinity')
+
+        if len(smiles_tensor) == 0:
+            return 0
+
+        _, protein_tensor = self.protein_to_numerical(
+            protein, encoder_uses_sequence=False
+        )
+
+        pred, pred_dict = self.affinity_predictor(
+            smiles_tensor, protein_tensor.repeat(smiles_tensor.shape[0], 1)
+        )
+
+        return np.squeeze(pred.detach().numpy())
+
+    def get_reward_paccmann(self, valid_smiles: List, cell_line: str) -> np.array:
+        """Get the reward from PaccMann.
+        
+        Args:
+            valid_smiles (List): A list of valid SMILES strings.
+            cell_line (str): Name of the cell line.
+        Returns:
+            np.array: computed reward (fixed to 1/(1+exp(x))).
+        """
+
+        smiles_tensor = self.smiles_to_numerical(valid_smiles, target='efficacy')
+
+        if len(smiles_tensor) == 0:
+            return 0
+
+        locations = ['gene_expression_' + str(x) for x in range(2128)]
+
+        gep_t = torch.Tensor(
+            self.gep_df.loc[self.gep_df['cell_line'] == cell_line  # yapf: disable
+                            ][locations].values
+        )
+
+        pred, pred_dict = self.efficacy_predictor(
+            smiles_tensor, gep_t.repeat(len(valid_smiles), 1)
+        )
+        log_micromolar_pred = self.get_log_molar(np.squeeze(pred.detach().numpy()))
+        lmps = [lmp if lmp < self.ic50_threshold else 10 for lmp in log_micromolar_pred]
+        return 1 / (1 + np.exp(lmps))
+
+    def policy_gradient(
+        self, cell_line: str, protein: str, epoch: int, batch_size: int = 10
+    ) -> Tuple:
+        """Implementation of the policy gradient algorithm.
+        
+        Args:
+            protein (str): Name of protein to generate drugs.
+            cell_line (str): Name of cell line to generate drugs.
+            epoch (int): training epoch.
+            batch_size (int, optional): Batch size. Defaults to 10.
+
+        Returns:
+            Tuple: Tuple containing the total reward and total loss.
+        """
+        rl_loss = 0
+        self.optimizer.zero_grad()
+
+        # get set
+        input_set, gep_t, protein_predictor = self.get_set(
+            cell_line, protein, batch_size, shuffle=True
+        )
+        # Encode the set
+        c_n, h_n, r_n = self.encoder(input_set)
```

Comment:
I was expecting the output of the set-AE-encoder to be a code (a fixed length-vector) and not a 3-tuple. Why you're using hidden and cell states here?

Reply:
yeah, this is because in the paper he concatenates all 3 to get the latent code, and says any combination of them would work. So here I just take the mean as my latent code.